test/unit/tools/test_tool_deps.py

```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
 
 
 @contextmanager
-def __environ(values, remove=[]):
+def __environ(values, keys_to_remove=[]):
     """
     Modify the environment for a test, adding/updating values in dict `values` and
     removing any environment variables mentioned in list `remove`.
     """
-    new_keys = set(environ.keys()) - set(values.keys())
+    new_keys = set(values.keys()) - set(environ.keys())
     old_environ = environ.copy()
     try:
         environ.update(values)
-        for to_remove in remove:
+        for key in keys_to_remove:
             try:
-                del environ[remove]
+                del environ[key]
+                new_keys.remove(key)  # Because key no longer exists
             except KeyError:
                 pass
         yield
     finally:
-        environ.update(old_environ)
         for key in new_keys:
             del environ[key]
+        environ.update(old_environ)
+
+
+def test_environ_contextmanager_env_restored():
+    """ os.environ must be preserved across calls to __environ """
+    key, val = 'foo_test_678363', '42'
+    os.environ[key] = val
+    with __environ({}, []):
+        pass
+    assert os.environ[key] == val
+    del os.environ[key]
+
+
+def test_environ_contextmanager_env_updated_and_restored():
+    """
+        os.environ must be preserved across calls to __environ
+        when updating existing or adding new key:value pairs
+    """
+    key, val = 'foo_test_678363', '42'
+    os.environ[key] = val
+    new_key, new_val1 = 'foo_test_new', '53'  # add new key + value
+    new_val2 = '43'  # update old key's value
+    to_update = {new_key: new_val1, key: new_val2}
+    with __environ(values=to_update):
+        pass
```

Comment:
What about testing that `os.environ` has been updated here?

Reply:
Done!